// next-translate/lib/cjs/plugin/loader.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var templateWithHoc_1 = __importDefault(require("./templateWithHoc"));
var templateWithLoader_1 = __importDefault(require("./templateWithLoader"));
var utils_1 = require("./utils");
function loader(rawCode) {
    var _a = this.query, hasGetInitialPropsOnAppJs = _a.hasGetInitialPropsOnAppJs, hasAppJs = _a.hasAppJs, extensionsRgx = _a.extensionsRgx, pagesPath = _a.pagesPath, hasLoadLocaleFrom = _a.hasLoadLocaleFrom;
    var normalizedPagesPath = pagesPath.replace(/\\/g, '/');
    var normalizedResourcePath = this.resourcePath.replace(/\\/g, '/');
    if (normalizedResourcePath.includes('node_modules/next/dist/pages/_app')) {
        if (hasAppJs)
            return rawCode;
        return utils_1.getDefaultAppJs(hasLoadLocaleFrom);
    }
    if (!normalizedResourcePath.startsWith(normalizedPagesPath))
        return rawCode;
    var page = normalizedResourcePath.replace(normalizedPagesPath, '/');
    var pageNoExt = page.replace(extensionsRgx, '');
    var code = rawCode.replace(utils_1.clearCommentsRgx, '');
    var typescript = page.endsWith('.ts') || page.endsWith('.tsx');
    if (!code.includes('export default'))
        return rawCode;
    if (hasGetInitialPropsOnAppJs) {
        return pageNoExt === '/_app'
            ? templateWithHoc_1.default(rawCode, { typescript: typescript, hasLoadLocaleFrom: hasLoadLocaleFrom })
            : rawCode;
    }
    if (pageNoExt === '/_app') {
        return templateWithHoc_1.default(rawCode, {
            skipInitialProps: true,
            typescript: typescript,
            hasLoadLocaleFrom: hasLoadLocaleFrom,
        });
    }
    if (utils_1.isPageToIgnore(page))
        return rawCode;
    var isWrapperWithExternalHOC = utils_1.hasHOC(code);
    var isDynamicPage = page.includes('[');
    var isGetInitialProps = !!code.match(/\WgetInitialProps\W/g);
    var isGetServerSideProps = utils_1.hasExportName(code, 'getServerSideProps');
    var isGetStaticPaths = utils_1.hasExportName(code, 'getStaticPaths');
    var isGetStaticProps = utils_1.hasExportName(code, 'getStaticProps');
    var hasLoader = isGetStaticProps || isGetServerSideProps || isGetInitialProps;
    if (isGetInitialProps || (!hasLoader && isWrapperWithExternalHOC)) {
        console.warn("\uD83D\uDEA8 [next-translate] In Next 10.x.x there is an issue related to i18n and getInitialProps. We recommend to replace getInitialProps to getServerSideProps on " + page + ". Issue: https://github.com/vercel/next.js/issues/18396");
        return templateWithHoc_1.default(rawCode, { typescript: typescript, hasLoadLocaleFrom: hasLoadLocaleFrom });
    }
    var loader = isGetServerSideProps || (!hasLoader && isDynamicPage && !isGetStaticPaths)
        ? 'getServerSideProps'
        : 'getStaticProps';
    return templateWithLoader_1.default(rawCode, {
        page: pageNoExt,
        typescript: typescript,
        loader: loader,
        hasLoader: hasLoader,
        hasLoadLocaleFrom: hasLoadLocaleFrom,
    });
}
exports.default = loader;
